pkg/storage/stores/tsdb/single_file_index.go

```diff
@@ -18,6 +18,8 @@ import (
 	util_log "github.com/grafana/loki/pkg/util/log"
 )
 
+var shouldCachePostings = false
```

Comment:
I don't think that we should use a module scope variable to define what postings client/reader should be used.

Reply:
I understand that it's pretty nasty to pass down the config to the constructor, though.

Reply:
It is an ugly workaround, but the other option is changing the signature everywhere to expect the cacheClient (but notice that the cacheClient would only be used here in the single_file_index). I'm fine with either of them tbh.

Reply:
I had a quick look at an approach to pass the config down to the constructor, and it seems it could be done using the opening function. Full diff:


```diff
diff --git a/pkg/storage/stores/tsdb/cached_postings_index.go b/pkg/storage/stores/tsdb/cached_postings_index.go
index 769f0a1b0..dc5c3baf6 100644
--- a/pkg/storage/stores/tsdb/cached_postings_index.go
+++ b/pkg/storage/stores/tsdb/cached_postings_index.go
@@ -24,15 +24,15 @@ type PostingsReader interface {
 
 var sharedCacheClient cache.Cache
 
-func NewCachedPostingsClient(reader IndexReader, logger log.Logger, cacheClient cache.Cache) PostingsReader {
-	return &cachedPostingsClient{
+func NewCachedPostingsReader(reader IndexReader, logger log.Logger, cacheClient cache.Cache) PostingsReader {
+	return &cachedPostingsReader{
 		reader:      reader,
 		cacheClient: cacheClient,
 		log:         logger,
 	}
 }
 
-type cachedPostingsClient struct {
+type cachedPostingsReader struct {
 	reader IndexReader
 
 	cacheClient cache.Cache
@@ -40,7 +40,7 @@ type cachedPostingsClient struct {
 	log log.Logger
 }
 
-func (c *cachedPostingsClient) ForPostings(ctx context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error {
+func (c *cachedPostingsReader) ForPostings(ctx context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error {
 	key := CanonicalLabelMatchersKey(matchers)
 	if postings, got := c.fetchPostings(ctx, key); got {
 		return fn(postings)
@@ -123,7 +123,7 @@ func encodedMatchersLen(matchers []*labels.Matcher) int {
 	return matchersLen
 }
 
-func (c *cachedPostingsClient) storePostings(ctx context.Context, postings index.Postings, canonicalMatchers string) error {
+func (c *cachedPostingsReader) storePostings(ctx context.Context, postings index.Postings, canonicalMatchers string) error {
 	dataToCache, err := diffVarintEncodeNoHeader(postings, 0)
 	if err != nil {
 		level.Warn(c.log).Log("msg", "couldn't encode postings", "err", err, "matchers", canonicalMatchers)
@@ -132,7 +132,7 @@ func (c *cachedPostingsClient) storePostings(ctx context.Context, postings index
 	return c.cacheClient.Store(ctx, []string{canonicalMatchers}, [][]byte{dataToCache})
 }
 
-func (c *cachedPostingsClient) fetchPostings(ctx context.Context, key string) (index.Postings, bool) {
+func (c *cachedPostingsReader) fetchPostings(ctx context.Context, key string) (index.Postings, bool) {
 	found, bufs, _, err := c.cacheClient.Fetch(ctx, []string{key})
 
 	if err != nil {
diff --git a/pkg/storage/stores/tsdb/cached_postings_index_test.go b/pkg/storage/stores/tsdb/cached_postings_index_test.go
index 85b8a6885..3bc2d97c8 100644
--- a/pkg/storage/stores/tsdb/cached_postings_index_test.go
+++ b/pkg/storage/stores/tsdb/cached_postings_index_test.go
@@ -78,7 +78,7 @@ func TestSingleIdxCached(t *testing.T) {
 					_, _ = head.Append(x.Labels, x.Labels.Hash(), x.Chunks)
 				}
 				reader := head.Index()
-				return NewTSDBIndex(reader, DefaultPostingsClient(reader))
+				return NewTSDBIndex(reader, NewPostingsReader(reader))
 			},
 		},
 	} {
diff --git a/pkg/storage/stores/tsdb/compactor.go b/pkg/storage/stores/tsdb/compactor.go
index 928252e02..1e9c1ff02 100644
--- a/pkg/storage/stores/tsdb/compactor.go
+++ b/pkg/storage/stores/tsdb/compactor.go
@@ -36,7 +36,8 @@ func (i indexProcessor) NewTableCompactor(ctx context.Context, commonIndexSet co
 }
 
 func (i indexProcessor) OpenCompactedIndexFile(ctx context.Context, path, tableName, userID, workingDir string, periodConfig config.PeriodConfig, logger log.Logger) (compactor.CompactedIndex, error) {
-	indexFile, err := OpenShippableTSDB(path)
+	opts := TSDBIndexOpts{UsePostingsCache: false}
+	indexFile, err := OpenShippableTSDB(path, opts)
 	if err != nil {
 		return nil, err
 	}
@@ -92,6 +93,8 @@ func (t *tableCompactor) CompactTable() error {
 	multiTenantIndices := make([]Index, len(multiTenantIndexes))
 	downloadPaths := make([]string, len(multiTenantIndexes))
 
+	opts := TSDBIndexOpts{UsePostingsCache: false}
+
 	// concurrently download and open all the multi-tenant indexes
 	err := concurrency.ForEachJob(t.ctx, len(multiTenantIndexes), readDBsConcurrency, func(ctx context.Context, job int) error {
 		downloadedAt, err := t.commonIndexSet.GetSourceFile(multiTenantIndexes[job])
@@ -100,7 +103,7 @@ func (t *tableCompactor) CompactTable() error {
 		}
 
 		downloadPaths[job] = downloadedAt
-		idx, err := OpenShippableTSDB(downloadedAt)
+		idx, err := OpenShippableTSDB(downloadedAt, opts)
 		if err != nil {
 			return err
 		}
@@ -205,6 +208,7 @@ func setupBuilder(ctx context.Context, userID string, sourceIndexSet compactor.I
 		}
 	}
 
+	opts := TSDBIndexOpts{UsePostingsCache: false}
 	// download all the existing compacted indexes and add them to the builder
 	for _, sourceIndex := range sourceIndexes {
 		path, err := sourceIndexSet.GetSourceFile(sourceIndex)
@@ -218,7 +222,7 @@ func setupBuilder(ctx context.Context, userID string, sourceIndexSet compactor.I
 			}
 		}()
 
-		indexFile, err := OpenShippableTSDB(path)
+		indexFile, err := OpenShippableTSDB(path, opts)
 		if err != nil {
 			return nil, err
 		}
@@ -387,7 +391,7 @@ func (c *compactedIndex) ToIndexFile() (index_shipper.Index, error) {
 		return nil, err
 	}
 
-	return NewShippableTSDBFile(id)
+	return NewShippableTSDBFile(id, TSDBIndexOpts{UsePostingsCache: false})
 }
 
 func getUnsafeBytes(s string) []byte {
diff --git a/pkg/storage/stores/tsdb/head_manager.go b/pkg/storage/stores/tsdb/head_manager.go
index c725cee4e..135fb573f 100644
--- a/pkg/storage/stores/tsdb/head_manager.go
+++ b/pkg/storage/stores/tsdb/head_manager.go
@@ -731,7 +731,7 @@ func (t *tenantHeads) tenantIndex(userID string, from, through model.Time) (idx
 	}
 
 	reader := tenant.indexRange(int64(from), int64(through))
-	idx = NewTSDBIndex(reader, &simplePostingsClient{reader: reader})
+	idx = NewTSDBIndex(reader, &simplePostingsReader{reader: reader})
 	if t.chunkFilter != nil {
 		idx.SetChunkFilterer(t.chunkFilter)
 	}
diff --git a/pkg/storage/stores/tsdb/manager.go b/pkg/storage/stores/tsdb/manager.go
index c5b970e46..7c685077e 100644
--- a/pkg/storage/stores/tsdb/manager.go
+++ b/pkg/storage/stores/tsdb/manager.go
@@ -129,7 +129,7 @@ func (m *tsdbManager) Start() (err error) {
 			indices++
 
 			prefixed := NewPrefixedIdentifier(id, filepath.Join(mulitenantDir, bucket), "")
-			loaded, err := NewShippableTSDBFile(prefixed)
+			loaded, err := NewShippableTSDBFile(prefixed, TSDBIndexOpts{UsePostingsCache: false})
 
 			if err != nil {
 				level.Warn(m.log).Log(
@@ -221,7 +221,7 @@ func (m *tsdbManager) buildFromHead(heads *tenantHeads, shipper indexshipper.Ind
 
 		level.Debug(m.log).Log("msg", "finished building tsdb for period", "pd", p, "dst", dst.Path(), "duration", time.Since(start))
 
-		loaded, err := NewShippableTSDBFile(dst)
+		loaded, err := NewShippableTSDBFile(dst, TSDBIndexOpts{UsePostingsCache: false})
 		if err != nil {
 			return err
 		}
diff --git a/pkg/storage/stores/tsdb/single_file_index.go b/pkg/storage/stores/tsdb/single_file_index.go
index 3d1e5b6bf..26412ffe5 100644
--- a/pkg/storage/stores/tsdb/single_file_index.go
+++ b/pkg/storage/stores/tsdb/single_file_index.go
@@ -27,17 +27,22 @@ var ErrAlreadyOnDesiredVersion = errors.New("tsdb file already on desired versio
 // GetRawFileReaderFunc returns an io.ReadSeeker for reading raw tsdb file from disk
 type GetRawFileReaderFunc func() (io.ReadSeeker, error)
 
-func OpenShippableTSDB(p string) (index_shipper.Index, error) {
+type TSDBIndexOpts struct {
+	UsePostingsCache bool
+}
+
+func OpenShippableTSDB(p string, opts TSDBIndexOpts) (index_shipper.Index, error) {
 	id, err := identifierFromPath(p)
 	if err != nil {
 		return nil, err
 	}
 
-	return NewShippableTSDBFile(id)
+	return NewShippableTSDBFile(id, opts)
 }
 
 func RebuildWithVersion(ctx context.Context, path string, desiredVer int) (index_shipper.Index, error) {
-	indexFile, err := OpenShippableTSDB(path)
+	opts := TSDBIndexOpts{UsePostingsCache: false}
+	indexFile, err := OpenShippableTSDB(path, opts)
 	if err != nil {
 		return nil, err
 	}
@@ -76,7 +81,7 @@ func RebuildWithVersion(ctx context.Context, path string, desiredVer int) (index
 	if err != nil {
 		return nil, err
 	}
-	return NewShippableTSDBFile(id)
+	return NewShippableTSDBFile(id, opts)
 }
 
 // nolint
@@ -92,8 +97,8 @@ type TSDBFile struct {
 	getRawFileReader GetRawFileReaderFunc
 }
 
-func NewShippableTSDBFile(id Identifier) (*TSDBFile, error) {
-	idx, getRawFileReader, err := NewTSDBIndexFromFile(id.Path())
+func NewShippableTSDBFile(id Identifier, opts TSDBIndexOpts) (*TSDBFile, error) {
+	idx, getRawFileReader, err := NewTSDBIndexFromFile(id.Path(), opts)
 	if err != nil {
 		return nil, err
 	}
@@ -125,42 +130,43 @@ type TSDBIndex struct {
 
 // Return the index as well as the underlying raw file reader which isn't exposed as an index
 // method but is helpful for building an io.reader for the index shipper
-func NewTSDBIndexFromFile(location string) (Index, GetRawFileReaderFunc, error) {
+func NewTSDBIndexFromFile(location string, opts TSDBIndexOpts) (Index, GetRawFileReaderFunc, error) {
 	reader, err := index.NewFileReader(location)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	tsdbIdx := NewTSDBIndex(reader, getPostingsClient(reader))
+	postingsReader := getPostingsReader(reader, opts.UsePostingsCache)
+	tsdbIdx := NewTSDBIndex(reader, postingsReader)
 
 	return tsdbIdx, func() (io.ReadSeeker, error) {
 		return reader.RawFileReader()
 	}, nil
 }
 
-func getPostingsClient(reader IndexReader) PostingsReader {
-	var postingsClient PostingsReader
+func getPostingsReader(reader IndexReader, usePostingsCache bool) PostingsReader {
+	var pr PostingsReader
 
-	if shouldCachePostings && sharedCacheClient != nil {
-		postingsClient = NewCachedPostingsClient(reader, util_log.Logger, sharedCacheClient)
+	if usePostingsCache && sharedCacheClient != nil {
+		pr = NewCachedPostingsReader(reader, util_log.Logger, sharedCacheClient)
 	}
 
-	if postingsClient == nil {
-		postingsClient = DefaultPostingsClient(reader)
+	if pr == nil {
+		pr = NewPostingsReader(reader)
 	}
 
-	return postingsClient
+	return pr
 }
 
-func DefaultPostingsClient(reader IndexReader) PostingsReader {
-	return &simplePostingsClient{reader: reader}
+func NewPostingsReader(reader IndexReader) PostingsReader {
+	return &simplePostingsReader{reader: reader}
 }
 
-type simplePostingsClient struct {
+type simplePostingsReader struct {
 	reader IndexReader
 }
 
-func (s *simplePostingsClient) ForPostings(_ context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error {
+func (s *simplePostingsReader) ForPostings(_ context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error {
 	p, err := PostingsForMatchers(s.reader, nil, matchers...)
 	if err != nil {
 		return err
diff --git a/pkg/storage/stores/tsdb/single_file_index_test.go b/pkg/storage/stores/tsdb/single_file_index_test.go
index f9c5988f3..ed8b4ae26 100644
--- a/pkg/storage/stores/tsdb/single_file_index_test.go
+++ b/pkg/storage/stores/tsdb/single_file_index_test.go
@@ -82,7 +82,7 @@ func TestSingleIdx(t *testing.T) {
 					_, _ = head.Append(x.Labels, x.Labels.Hash(), x.Chunks)
 				}
 				reader := head.Index()
-				return NewTSDBIndex(reader, DefaultPostingsClient(reader))
+				return NewTSDBIndex(reader, NewPostingsReader(reader))
 			},
 		},
 	} {
diff --git a/pkg/storage/stores/tsdb/store.go b/pkg/storage/stores/tsdb/store.go
index d848216c9..079a77b41 100644
--- a/pkg/storage/stores/tsdb/store.go
+++ b/pkg/storage/stores/tsdb/store.go
@@ -21,6 +21,7 @@ import (
 	"github.com/grafana/loki/pkg/storage/stores/index"
 	"github.com/grafana/loki/pkg/storage/stores/indexshipper"
 	"github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads"
+	indexshipper_index "github.com/grafana/loki/pkg/storage/stores/indexshipper/index"
 	tsdb_index "github.com/grafana/loki/pkg/storage/stores/tsdb/index"
 )
 
@@ -32,6 +33,7 @@ type store struct {
 	index.Reader
 	indexShipper      indexshipper.IndexShipper
 	indexWriter       IndexWriter
+	indexReadCache    cache.Cache
 	backupIndexWriter index.Writer
 	logger            log.Logger
 	stopOnce          sync.Once
@@ -62,6 +64,7 @@ func NewStore(
 	storeInstance := &store{
 		backupIndexWriter: backupIndexWriter,
 		logger:            logger,
+		indexReadCache:    idxCache,
 	}
 
 	if err := storeInstance.init(name, indexShipperCfg, schemaCfg, objectClient, limits, tableRange, reg, idxCache); err != nil {
@@ -74,10 +77,9 @@ func NewStore(
 func (s *store) init(name string, indexCfg IndexCfg, schemaCfg config.SchemaConfig, objectClient client.ObjectClient,
 	limits downloads.Limits, tableRange config.TableRange, reg prometheus.Registerer, idxCache cache.Cache) error {
 
-	sharedCacheClient = idxCache
-
-	if indexCfg.CachePostings {
-		shouldCachePostings = true
+	usePostingsCache := indexCfg.Mode == indexshipper.ModeReadOnly && idxCache != nil
+	openFn := func(p string) (indexshipper_index.Index, error) {
+		return OpenShippableTSDB(p, TSDBIndexOpts{UsePostingsCache: usePostingsCache})
 	}
 
 	var err error
@@ -86,7 +88,7 @@ func (s *store) init(name string, indexCfg IndexCfg, schemaCfg config.SchemaConf
 		objectClient,
 		limits,
 		nil,
-		OpenShippableTSDB,
+		openFn,
 		tableRange,
 		prometheus.WrapRegistererWithPrefix("loki_tsdb_shipper_", reg),
 		s.logger,
diff --git a/tools/tsdb/tsdb-map/main_test.go b/tools/tsdb/tsdb-map/main_test.go
index ee12cf720..4d4908529 100644
--- a/tools/tsdb/tsdb-map/main_test.go
+++ b/tools/tsdb/tsdb-map/main_test.go
@@ -93,7 +93,7 @@ func BenchmarkQuery_GetChunkRefs(b *testing.B) {
 		if err != nil {
 			panic(err)
 		}
-		idx := tsdb.NewTSDBIndex(reader, tsdb.DefaultPostingsClient(reader))
+		idx := tsdb.NewTSDBIndex(reader, tsdb.NewPostingsReader(reader))
 		b.Run(bm.name, func(b *testing.B) {
 			refs := tsdb.ChunkRefsPool.Get()
 			for i := 0; i < b.N; i++ {
@@ -118,7 +118,7 @@ func BenchmarkQuery_GetChunkRefsSharded(b *testing.B) {
 		if err != nil {
 			panic(err)
 		}
-		idx := tsdb.NewTSDBIndex(reader, tsdb.DefaultPostingsClient(reader))
+		idx := tsdb.NewTSDBIndex(reader, tsdb.NewPostingsReader(reader))
 		shardFactor := 16
 
 		b.Run(bm.name, func(b *testing.B) {
```